### PR TITLE
Make Scipy more robust & faster

### PR DESCRIPTION
**Pull request by st--**
* Adds an explicit check to Scipy() that all elements of `variables` are actually tf.Variable instances (unconstrained variables) - accidentally passing `model.trainable_parameters` may work but give the wrong answers for parameters with transforms! Passing `model.trainable_variables` is the right thing to do.
* Updates _compute_loss_and_gradients to only compute gradients wrt passed-in variables.
* Fix mixture density network notebook, which had erroneously passed model.trainable_parameters instead of model.trainable_variables.

### CODE REVIEW

File: doc/source/notebooks/tailor/mixture_density_network.pct.py [hunk 1]
```diff
@@ -182,7 +182,7 @@ def log_marginal_likelihood(self, data: Tuple[tf.Tensor, tf.Tensor]):
 
 Scipy().minimize(
     tf.function(lambda: -model.log_marginal_likelihood(data)),
-    variables=model.trainable_parameters,
+    variables=model.trainable_variables,
```
Comment (st--) [inline doc/source/notebooks/tailor/mixture_density_network.pct.py]: this might've been a bug? :thinking: @vdutor

File: gpflow/optimizers/scipy.py [hunk 2]
```diff
@@ -134,8 +138,9 @@ def assign_tensors(to_tensors: Iterable[tf.Variable], values: Iterable[tf.Tensor
...
 
-def _compute_loss_and_gradients(loss_cb: LossClosure, variables: V) -> Tuple[tf.Tensor, V]:
-    with tf.GradientTape() as tape:
-        loss = loss_cb()
-    grads = tape.gradient(loss, variables)
+def _compute_loss_and_gradients(loss_closure: LossClosure, variables: V) -> Tuple[tf.Tensor, V]:
+    with tf.GradientTape(watch_accessed_variables=False) as tape:
+        tape.watch(variables)
+        loss = loss_closure()
+        grads = tape.gradient(loss, variables)
```
Comment (awav) [inline gpflow/optimizers/scipy.py]:
```suggestion
    grads = tape.gradient(loss, variables)
```

This is not a speed-up, but slow-down ;)

File: gpflow/optimizers/scipy.py [hunk 3]
```diff
@@ -52,8 +52,12 @@ def minimize(
             object. See the Scipy documentation for description of attributes.
         """
         if not callable(closure):
-            raise TypeError("Callable object expected.")  # pragma: no cover
+            raise TypeError("`closure` is expected to be a callable object.")  # pragma: no cover
```
Comment (awav) [inline gpflow/optimizers/scipy.py]:
```suggestion
            raise TypeError("Closure argument is expected to be a callable object.")  # pragma: no cover
```

File: doc/source/notebooks/tailor/mixture_density_network.pct.py [hunk 4]
```diff
@@ -182,7 +182,7 @@ def log_marginal_likelihood(self, data: Tuple[tf.Tensor, tf.Tensor]):
 
 Scipy().minimize(
     tf.function(lambda: -model.log_marginal_likelihood(data)),
-    variables=model.trainable_parameters,
+    model.trainable_variables,
```
Comment (awav) [inline doc/source/notebooks/tailor/mixture_density_network.pct.py]: Technically, this is okay-ish thing to do.

File: gpflow/optimizers/scipy.py [hunk 5]
```diff
@@ -52,8 +52,12 @@ def minimize(
...
         variables = tuple(variables)
+        if not all(isinstance(v, tf.Variable) for v in variables):
+            raise TypeError(
+                "`variables` is expected to only contain tf.Variable instances (use model.trainable_variables, not model.trainable_parameters)"
+            )
```
Comment (st--) [inline gpflow/optimizers/scipy.py]:
```suggestion
            )  # pragma: no cover
```
